File: source/blender/editors/space_buttons/buttons_object.c
/**
 * $Id$
 *
 * ***** BEGIN GPL LICENSE BLOCK *****
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * The Original Code is Copyright (C) 2001-2002 by NaN Holding BV.
 * All rights reserved.
 *
 * The Original Code is: all of this file.
 *
 * Contributor(s): none yet.
 *
 * ***** END GPL LICENSE BLOCK *****
 */

#include <stdio.h>
#include <stdlib.h>

#include "MEM_guardedalloc.h"

#include "DNA_group_types.h"
#include "DNA_object_types.h"
#include "DNA_screen_types.h"
#include "DNA_space_types.h"

#include "BLI_listbase.h"

#include "BKE_context.h"
#include "BKE_group.h"
#include "BKE_main.h"
#include "BKE_screen.h"

#include "UI_interface.h"
#include "UI_resources.h"

#include "RNA_access.h"

#include "WM_types.h"

#if 0
static void object_panel_transform(const bContext *C, Panel *pnl)
{
	uiLayout *layout= pnl->layout;
	Object *ob= CTX_data_active_object(C);
	PointerRNA obptr;

	RNA_id_pointer_create(&ob->id, &obptr);

	uiTemplateColumnFlow(layout, 3);
	uiItemR(layout, NULL, 0, &obptr, "location");
	uiItemR(layout, NULL, 0, &obptr, "rotation");
	uiItemR(layout, NULL, 0, &obptr, "scale");
}

static void object_panel_groups(const bContext *C, Panel *pnl)
{
	uiLayout *layout= pnl->layout;
	Main *bmain= CTX_data_main(C);
	Object *ob= CTX_data_active_object(C);
	Group *group;
	PointerRNA obptr, groupptr;
	uiLayout *sublayout;

	RNA_id_pointer_create(&ob->id, &obptr);

	uiTemplateColumnFlow(layout, 2);
	uiItemR(layout, NULL, 0, &obptr, "pass_index");
	uiItemR(layout, NULL, 0, &obptr, "parent");

	/* uiTemplateLeftRight(layout);
	uiItemO(layout, NULL, 0, "OBJECT_OT_add_group"); */

	for(group=bmain->group.first; group; group=group->id.next) {
		if(object_in_group(ob, group)) {
			RNA_id_pointer_create(&group->id, &groupptr);

			sublayout= uiTemplateStack(layout);

			uiTemplateLeftRight(sublayout);
			uiTemplateSlot(sublayout, UI_TSLOT_LR_LEFT);
			uiItemR(sublayout, NULL, 0, &groupptr, "name");
			// uiTemplateSlot(sublayout, UI_TSLOT_RIGHT);
			// uiItemO(sublayout, "", ICON_X, "OBJECT_OT_remove_group");

			uiTemplateColumnFlow(sublayout, 2);
			uiItemR(sublayout, NULL, 0, &groupptr, "layer");
			uiItemR(sublayout, NULL, 0, &groupptr, "dupli_offset");
		}
	}
}

static void object_panel_display(const bContext *C, Panel *pnl)
{
	uiLayout *layout= pnl->layout;
	Object *ob= CTX_data_active_object(C);
	PointerRNA obptr;

	RNA_id_pointer_create(&ob->id, &obptr);

	uiTemplateColumnFlow(layout, 2);
	uiItemR(layout, "Type", 0, &obptr, "max_draw_type");
	uiItemR(layout, "Bounds", 0, &obptr, "draw_bounds_type");

	uiTemplateColumnFlow(layout , 2);
	uiItemR(layout, "Name", 0, &obptr, "draw_name");
	uiItemR(layout, "Axis", 0, &obptr, "draw_axis");
	uiItemR(layout, "Wire", 0, &obptr, "draw_wire");
	uiItemR(layout, "Texture Space", 0, &obptr, "draw_texture_space");
	uiItemR(layout, "X-Ray", 0, &obptr, "x_ray");
	uiItemR(layout, "Transparency", 0, &obptr, "draw_transparent");
}

static void object_panel_duplication(const bContext *C, Panel *pnl)
{
	uiLayout *layout= pnl->layout;
	Object *ob= CTX_data_active_object(C);
	PointerRNA obptr;

	RNA_id_pointer_create(&ob->id, &obptr);

	uiTemplateColumn(layout);
	uiItemR(layout, "", 0, &obptr, "dupli_type");

	if(RNA_enum_get(&obptr, "dupli_type") == OB_DUPLIFRAMES) {
		uiTemplateColumnFlow(layout, 2);
		uiItemR(layout, "Start:", 0, &obptr, "dupli_frames_start");
		uiItemR(layout, "End:", 0, &obptr, "dupli_frames_end");

		uiItemR(layout, "On:", 0, &obptr, "dupli_frames_on");
		uiItemR(layout, "Off:", 0, &obptr, "dupli_frames_off");
	}
}

static void object_panel_animation(const bContext *C, Panel *pnl)
{
	uiLayout *layout= pnl->layout;
	Object *ob= CTX_data_active_object(C);
	PointerRNA obptr;

	RNA_id_pointer_create(&ob->id, &obptr);

	uiTemplateColumn(layout);
	uiTemplateSlot(layout, UI_TSLOT_COLUMN_1);
	uiItemL(layout, "Time Offset:", 0);
	uiItemR(layout, "Edit", 0, &obptr, "time_offset_edit");
	uiItemR(layout, "Particle", 0, &obptr, "time_offset_particle");
	uiItemR(layout, "Parent", 0, &obptr, "time_offset_parent");
	uiItemR(layout, NULL, 0, &obptr, "slow_parent");
	uiItemR(layout, "Offset: ", 0, &obptr, "time_offset");
	uiTemplateSlot(layout, UI_TSLOT_COLUMN_2);
	uiItemL(layout, "Tracking:", 0);
	uiItemR(layout, "Axis: ", 0, &obptr, "track_axis");
	uiItemR(layout, "Up Axis: ", 0, &obptr, "up_axis");
	uiItemR(layout, "Rotation", 0, &obptr, "track_rotation");
}
#endif

void buttons_object_register(ARegionType *art)
{
#if 0
	PanelType *pt;

	/* panels: transform */
	pt= MEM_callocN(sizeof(PanelType), "spacetype buttons panel");
	pt->idname= "OBJECT_PT_transform";
	pt->name= "Transform";
	pt->context= "object";
	pt->draw= object_panel_transform;
	BLI_addtail(&art->paneltypes, pt);

	/* panels: groups */
	pt= MEM_callocN(sizeof(PanelType), "spacetype buttons panel");
	pt->idname= "OBJECT_PT_groups";
	pt->name= "Groups";
	pt->context= "object";
	pt->draw= object_panel_groups;
	BLI_addtail(&art->paneltypes, pt);

	/* panels: display */
	pt= MEM_callocN(sizeof(PanelType), "spacetype buttons panel");
	pt->idname= "OBJECT_PT_display";
	pt->name= "Display";
	pt->context= "object";
	pt->draw= object_panel_display;
	BLI_addtail(&art->paneltypes, pt);

	/* panels: duplication */
	pt= MEM_callocN(sizeof(PanelType), "spacetype buttons panel");
	pt->idname= "OBJECT_PT_duplication";
	pt->name= "Duplication";
	pt->context= "object";
	pt->draw= object_panel_duplication;
	BLI_addtail(&art->paneltypes, pt);

	/* panels: animation */
	pt= MEM_callocN(sizeof(PanelType), "spacetype buttons panel");
	pt->idname= "OBJECT_PT_animation";
	pt->name= "Animation";
	pt->context= "object";
	pt->draw= object_panel_animation;
	BLI_addtail(&art->paneltypes, pt);
#endif
}

File: release/scripts/ui/buttons_object.py

import bpy

class ObjectButtonsPanel(bpy.types.Panel):
	__space_type__ = 'PROPERTIES'
	__region_type__ = 'WINDOW'
	__context__ = "object"

class OBJECT_PT_context_object(ObjectButtonsPanel):
	__show_header__ = False

	def draw(self, context):
		layout = self.layout
		
		ob = context.object
		
		row = layout.row()
		row.itemL(text="", icon='ICON_OBJECT_DATA')
		row.itemR(ob, "name", text="")

class OBJECT_PT_transform(ObjectButtonsPanel):
	__label__ = "Transform"

	def draw(self, context):
		layout = self.layout
		
		ob = context.object
		
		layout.itemR(ob, "rotation_mode")

		row = layout.row()
		
		row.column().itemR(ob, "location")
		if ob.rotation_mode == 'QUATERNION':
			row.column().itemR(ob, "rotation_quaternion", text="Rotation")
		elif ob.rotation_mode == 'AXIS_ANGLE':
			#row.column().itemL(text="Rotation")
			#row.column().itemR(pchan, "rotation_angle", text="Angle")
			#row.column().itemR(pchan, "rotation_axis", text="Axis")
			row.column().itemR(ob, "rotation_axis_angle", text="Rotation")
		else:
			row.column().itemR(ob, "rotation_euler", text="Rotation")
			
		row.column().itemR(ob, "scale")
		
class OBJECT_PT_transform_locks(ObjectButtonsPanel):
	__label__ = "Transform Locks"
	__default_closed__ = True
	
	def draw(self, context):
		layout = self.layout
		
		ob = context.object
		
		row = layout.row()
		
		col = row.column()
		col.itemR(ob, "lock_location")
		
		col = row.column()
		if ob.rotation_mode in ('QUATERNION', 'AXIS_ANGLE'):
			col.itemR(ob, "lock_rotations_4d", text="Lock Rotation")
			if ob.lock_rotations_4d:
				col.itemR(ob, "lock_rotation_w", text="W")
			col.itemR(ob, "lock_rotation", text="")
		else:
			col.itemR(ob, "lock_rotation", text="Rotation")
		
		row.column().itemR(ob, "lock_scale")

class OBJECT_PT_relations(ObjectButtonsPanel):
	__label__ = "Relations"

	def draw(self, context):
		layout = self.layout
		
		ob = context.object

		split = layout.split()
		
		col = split.column()
		col.itemR(ob, "layers")
		col.itemS()
		col.itemR(ob, "pass_index")

		col = split.column()
		col.itemL(text="Parent:")
		col.itemR(ob, "parent", text="")

		sub = col.column()
		split = sub.split(percentage=0.3)
		split.itemL(text="Type:")
		split.itemR(ob, "parent_type", text="")
		parent = ob.parent
		if parent and ob.parent_type == 'BONE' and parent.type == 'ARMATURE':
			sub.item_pointerR(ob, "parent_bone", parent.data, "bones", text="")
		sub.active = parent != None

class OBJECT_PT_groups(ObjectButtonsPanel):
	__label__ = "Groups"

	def draw(self, context):
		layout = self.layout
		
		ob = context.object

		split = layout.split()
		split.item_menu_enumO("object.group_add", "group", text="Add to Group")
		split.itemL()

		for group in bpy.data.groups:
			if ob.name in group.objects:
				col = layout.column(align=True)

				col.set_context_pointer("group", group)

				row = col.box().row()
				row.itemR(group, "name", text="")
				row.itemO("object.group_remove", text="", icon='VICON_X')

				split = col.box().split()
				split.column().itemR(group, "layer", text="Dupli")
				split.column().itemR(group, "dupli_offset", text="")

class OBJECT_PT_display(ObjectButtonsPanel):
	__label__ = "Display"

	def draw(self, context):
		layout = self.layout
		
		ob = context.object
		
		split = layout.split()
		col = split.column()
		col.itemR(ob, "max_draw_type", text="Type")
		col = split.column()
		row = col.row()
		row.itemR(ob, "draw_bounds", text="Bounds")
		row.itemR(ob, "draw_bounds_type", text="")

		flow = layout.column_flow()
		flow.itemR(ob, "draw_name", text="Name")
		flow.itemR(ob, "draw_axis", text="Axis")
		flow.itemR(ob, "draw_wire", text="Wire")
		flow.itemR(ob, "draw_texture_space", text="Texture Space")
		flow.itemR(ob, "x_ray", text="X-Ray")
		flow.itemR(ob, "draw_transparent", text="Transparency")

class OBJECT_PT_duplication(ObjectButtonsPanel):
	__label__ = "Duplication"

	def draw(self, context):
		layout = self.layout
		
		ob = context.object

		layout.itemR(ob, "dupli_type", expand=True)

		if ob.dupli_type == 'FRAMES':
			split = layout.split()
			
			col = split.column(align=True)
			col.itemR(ob, "dupli_frames_start", text="Start")
			col.itemR(ob, "dupli_frames_end", text="End")
			
			col = split.column(align=True)
			col.itemR(ob, "dupli_frames_on", text="On")
			col.itemR(ob, "dupli_frames_off", text="Off")
			
			layout.itemR(ob, "dupli_frames_no_speed", text="No Speed")

		elif ob.dupli_type == 'VERTS':
			layout.itemR(ob, "dupli_verts_rotation", text="Rotation")

		elif ob.dupli_type == 'FACES':
			row = layout.row()
			row.itemR(ob, "dupli_faces_scale", text="Scale")
			row.itemR(ob, "dupli_faces_inherit_scale", text="Inherit Scale")

		elif ob.dupli_type == 'GROUP':
			layout.itemR(ob, "dupli_group", text="Group")

class OBJECT_PT_animation(ObjectButtonsPanel):
	__label__ = "Animation"

	def draw(self, context):
		layout = self.layout
		
		ob = context.object
		
		split = layout.split()
		
		col = split.column()
		col.itemL(text="Time Offset:")
		col.itemR(ob, "time_offset_edit", text="Edit")
		row = col.row()
		row.itemR(ob, "time_offset_particle", text="Particle")
		row.active = len(ob.particle_systems) != 0
		row = col.row()
		row.itemR(ob, "time_offset_parent", text="Parent")
		row.active = ob.parent != None
		row = col.row()
		row.itemR(ob, "slow_parent")
		row.active = ob.parent != None
		col.itemR(ob, "time_offset", text="Offset")

		col = split.column()
		col.itemL(text="Track:")
		col.itemR(ob, "track", text="")
		col.itemR(ob, "track_axis", text="Axis")
		col.itemR(ob, "up_axis", text="Up Axis")
		row = col.row()
		row.itemR(ob, "track_override_parent", text="Override Parent")
		row.active = ob.parent != None

bpy.types.register(OBJECT_PT_context_object)
bpy.types.register(OBJECT_PT_transform)
bpy.types.register(OBJECT_PT_transform_locks)
bpy.types.register(OBJECT_PT_relations)
bpy.types.register(OBJECT_PT_groups)
bpy.types.register(OBJECT_PT_display)
bpy.types.register(OBJECT_PT_duplication)
bpy.types.register(OBJECT_PT_animation)
